### PR TITLE
Fix ADLSG2 intermittent failure

### PR DESCRIPTION
**Pull request by adreed-msft**
When a ADLSG2 folder already exists with blobs under it and we attempt to overwrite the properties on it, the transfer will fail. As a result, we opt for a best-effort overwrite.

### CODE REVIEW

**Comment by gapra-msft** [comment]
@adreed-msft was this issue reported on Github? If so can we link it here?

**Comment by adreed-msft** [comment]
No-- This was detected through #1898.

**Comment by gapra-msft** [comment]
Can we write a test for this?

**Comment by tasherif-msft** [comment]
Let's add some tests and get prioritize it to get merged in to get our pipelines stable

**Review by gapra-msft — CHANGES_REQUESTED** [review]
TODO: Tests

**Review by siminsavani-msft — APPROVED** [review]

File: ste/sender-blobFolders.go [hunk 1]
```diff
@@ -53,6 +54,53 @@ func newBlobFolderSender(jptm IJobPartTransferMgr, destination string, p pipelin
 	return out, nil
 }
 
+func (b *blobFolderSender) overwriteDFSProperties() (string, error) {
+	b.jptm.Log(pipeline.LogWarning, "It is impossible to completely overwrite a folder with existing content under it on a hierarchical namespace storage account. A best-effort attempt will be made, but if CPK does not match the transfer will fail.")
+
+	b.metadataToApply["hdi_isfolder"] = "true" // Set folder metadata flag
```
Comment (vibhansa-msft) [inline ste/sender-blobFolders.go:60]: Is this function generic as name does not suggest this is only for folders though we are setting hdi_isfolder unconditionally here.

File: ste/sender-blobFolders.go [hunk 2]
```diff
@@ -103,6 +162,26 @@ func (b *blobFolderSender) EnsureFolderExists() error {
 		return fmt.Errorf("when creating folder: %w", err)
 	}
 
+	// Upload ADLS Gen 2 ACLs
+	if b.jptm.FromTo() == common.EFromTo.BlobBlob() && b.jptm.Info().PreserveSMBPermissions.IsTruthy() {
```
Comment (vibhansa-msft) [inline ste/sender-blobFolders.go:166]: Add test cases for both the changes.

**Review by nakulkar-msft — APPROVED** [review]

File: ste/sender-blobFolders.go (hunk 1 above)
Comment (gapra-msft) [reply ste/sender-blobFolders.go:60]: @vibhansa-msft The type is a blobFolderSender so I think its safe to always add the metadata

File: ste/sender-blobFolders.go [hunk 3]
```diff
@@ -53,6 +54,53 @@ func newBlobFolderSender(jptm IJobPartTransferMgr, destination string, p pipelin
...
+	if err != nil {
+		return "Set Blob Tags", err
+	}
+	_, err = b.destination.SetHTTPHeaders(b.jptm.Context(), b.headersToAppply, azblob.BlobAccessConditions{})
+	if err != nil {
+		return "Set HTTP Headers", err
+	}
+
+	// Upload ADLS Gen 2 ACLs
+	if b.jptm.FromTo() == common.EFromTo.BlobBlob() && b.jptm.Info().PreserveSMBPermissions.IsTruthy() {
```
Comment (gapra-msft) [inline ste/sender-blobFolders.go:82]: how do we know if the destination is adls?

File: ste/sender-blobFolders.go [hunk 4]
```diff
@@ -103,6 +162,26 @@ func (b *blobFolderSender) EnsureFolderExists() error {
 		return fmt.Errorf("when creating folder: %w", err)
 	}
 
+	// Upload ADLS Gen 2 ACLs
+	if b.jptm.FromTo() == common.EFromTo.BlobBlob() && b.jptm.Info().PreserveSMBPermissions.IsTruthy() {
+		bURLParts := azblob.NewBlobURLParts(b.destination.URL())
```
Comment (gapra-msft) [inline ste/sender-blobFolders.go:167]: this code seems to be duplicate of above, should we make this a method?

**Review by gapra-msft — APPROVED** [review]

**Review by vibhansa-msft — APPROVED** [review]

File: ste/sender-blobFolders.go (hunk 3 above)
Comment (adreed-msft) [reply ste/sender-blobFolders.go:82]: The only place this is called is when we receive an HNS specific error message.

**Review by tasherif-msft — APPROVED** [review]

**Pull request merged**